From e9a06a5f6300c95fb4bc4e376298bbf3ccd4639e Mon Sep 17 00:00:00 2001
From: Ross Lagerwall <ross.lagerwall@citrix.com>
Date: Tue, 26 Apr 2016 13:59:38 +0100
Subject: [PATCH] efi: Workaround page fault during runtime service

Some hardware makes use of memory of type EfiACPIMemoryNVS during the
ResetSystem runtime service but does not mark the memory as needing a
runtime mapping causing a fatal page fault.

To workaround this, map this type of memory for runtime services
regardless of whether it marked with EFI_MEMORY_RUNTIME.

The hardware on which this behavior was observed:
Dell OptiPlex 9020
Firmware version: A15
Firmware release date: 11/08/2015

Signed-off-by: Ross Lagerwall <ross.lagerwall@citrix.com>
diff --git a/xen/common/efi/boot.c b/xen/common/efi/boot.c
index 577cd6e..1fda1e8 100644
--- a/xen/common/efi/boot.c
+++ b/xen/common/efi/boot.c
@@ -140,6 +140,7 @@ static SIMPLE_TEXT_OUTPUT_INTERFACE *__initdata StdErr;
 static UINT32 __initdata mdesc_ver;
 static bool __initdata map_bs;
 static bool __initdata map_res = true;
+static bool __initdata map_nvs = true;
 
 static struct file __initdata cfg;
 static struct file __initdata kernel;
@@ -1420,6 +1421,8 @@ void __init efi_init_memory(void)
               (!map_bs ||
                (desc->Type != EfiBootServicesCode &&
                 desc->Type != EfiBootServicesData)) &&
+              (!map_nvs ||
+               desc->Type != EfiACPIMemoryNVS) &&
               (!map_res ||
                desc->Type != EfiReservedMemoryType)) )
             continue;
